### ISSUE TITLE
R memory limitation while using loadSQM

### ISSUE THREAD

**Issue by yjiakang** [jtamames/SqueezeMeta#198]
Hello, I have completed  36 pair-end metagenomic sample using your pipeline with the `sequential` mode individually. Then I want to combine them use the `combineSQMlite` function. And my computer broke down due to limited memory. My computer has 16Gb RAM. Is this normal? Here is my code: 
```
library(SQMtools)
library(parallel)
setwd("D:/phd/tomato_micr/meta/SqueezeMeta")
samples <- c("A57RS1T1", "A57RS1T2", "A57RS1T2C", "A57RS2T1", 
             "A57RS2T2", "A57RS2T2C", "A57RS3T1", "A57RS3T2", 
             "A57RS3T2C", "HF12RS1T1", "HF12RS1T2", "HF12RS1T2C", 
             "HF12RS2T1", "HF12RS2T2", "HF12RS2T2C", "HF12RS3T1", 
             "HF12RS3T2", "HF12RS3T2C", "HG64RS1T1", "HG64RS1T2", 
             "HG64RS1T2C", "HG64RS2T1", "HG64RS2T2", "HG64RS2T2C",
             "HG64RS3T1", "HG64RS3T2", "HG64RS3T2C", "HG70RS1T1", 
             "HG70RS1T2", "HG70RS1T2C", "HG70RS2T1", "HG70RS2T2", 
             "HG70RS2T2C", "HG70RS3T1", "HG70RS3T2", "HG70RS3T2C")
cl = makeCluster(2)
memory.limit(16000000)
SQMobjs = parLapply(cl, samples, loadSQM)
```

**Comment by fpusan** [comment]
It could happen. `loadSQM` loads everything into memory, including all the sequences, annotations and abundances for each ORF, etc. If your projects are large and there are 36 of them, they could very well use more than 16Gb.

An alternative is to use the function `loadSQMlite` over the `results/tables` directory of each project. This will only load the functional and taxonomic profiles.

**Comment by yjiakang** [comment]
The loadSQMlite perfectly worked. However, there are a large proportion of Unmapped in the taxonomy. I checked  the 22.project.stats file, is there any clue?
```
#------------------- Statistics on ORFs
#       Assembly        A57RS1T1
Number of ORFs  2330258 2325463
Number of rRNAs 3417    3405
Number of tRNAs/tmRNAs  10531   10495
ORFs by Aragorn 10531   10495
ORFs by Prodigal        2316310 2311563
ORFs by barrnap 3417    3405
Orphans (no hits)       385504  383895
No tax assigned (with hits)     10530   10521
KEGG annotations        1254598 1252369
COG annotations 1668013 1665305
Pfam annotations        916511  915716
```

**Comment by jtamames** [comment]
Hello

You mean the orphans? In my opinion, this is a very nice annotation, with just 16% orphans. It is usual to have higher ratios of orphans, see for instance #157 
Best,
J

**Comment by fpusan** [comment]
What is the actual proportion of unmapped reads in the taxonomy?

**Comment by yjiakang** [comment]
Hello, there are ~ 50% unmapped reads in the taxonomy. Please see the attached picture. Thanks for your attention. 
![phylum](https://user-images.githubusercontent.com/49177016/100222307-ddcead80-2f54-11eb-8efc-d3fabd10d008.png)

**Comment by jtamames** [comment]
Could you paste your 10*mappingstat file?

Thanks!

**Comment by yjiakang** [comment]
[10.A57RS1T1.mappingstat.zip](https://github.com/jtamames/SqueezeMeta/files/5596420/10.A57RS1T1.mappingstat.zip)

**Comment by yjiakang** [comment]
By the way, my commond is 
```
SqueezeMeta.pl -m sequential -s /public/home/jkyin/yjk/tomato_micro/sqm_one_by_one/sample_file/A57RS1T1 -t 12 -f /public/home/jkyin/yjk/tomato_micro/raw -extassembly /public/home/jkyin/yjk/tomato_micro/squeezemeta/four_cultivars/intermediate/01.four_cultivars.A57RS1T1.fasta
```

**Comment by yjiakang** [comment]
By the way, my command is
```
SqueezeMeta.pl -m sequential -s /public/home/jkyin/yjk/tomato_micro/sqm_one_by_one/sample_file/A57RS1T1 -t 12 -f /public/home/jkyin/yjk/tomato_micro/raw -extassembly /public/home/jkyin/yjk/tomato_micro/squeezemeta/four_cultivars/intermediate/01.four_cultivars.A57RS1T1.fasta
```

**Comment by jtamames** [comment]
Thanks for the file. Yep, the mapping percentage is probably too low (<50%). The most likely explanation for this is that the sample is very rich in species and the sequencing was not enough to get a good coverage for many of them. As a consequence, they don´t assemble in contigs.  In these cases I suggest using sqm_reads.pl or sqm_longreads.pl, that is, skipping the assembly and working on reads directly

**Comment by yjiakang** [comment]
Thanks for reply.
Is there any problem if I filter the unmapped? Does this cause misundestanding or bias of the results?
Best

**Comment by fpusan** [comment]
It is not ideal, as it shows that your assembly is not capturing the whole diversity of genomes in your samples. In your case all the samples have around the same percentage of unmapped reads, so I would not expect large biases to be happening (but take my works with a large pinch of salt, as I am no statistician).

For sure I would compare the results with those obtained with `sqm_reads.pl`, this should give you a hint on what the missing taxa might be.

Also, I see that you provided the assembly using `--extassembly`. Is there a chance that you provided the wrong assembly for those samples?

**Comment by yjiakang** [comment]
Yep, I checked the file following the `--extassembly` tag,  and I think it is right. It is located in the `intermediate` directory that was produced by SqueezeMeta.pl with merged mode. 

I am trying  the sqm_reads.pl  with the following command 
```
sqm_reads.pl -p A57RS1T1 -s /public/home/jkyin/yjk/tomato_micro/sqm_one_by_one/sample_file/A57RS1T1 -t 8 -f /public/home/jkyin/yjk/tomato_micro/raw
```
I will tell you the difference between the results produced by `sqm_reads.pl` and `SqueezeMeta.pl` after run. 

I have another question: is the results associated with the function part trustworthy?
Thanks!

**Comment by fpusan** [comment]
We of course make an effort to retrieve trustworthy results and improve the algorithms when we detect inconsistencies (we use SqueezeMeta for our own research), but functional annotation can be tricky. Please see the discussion in #187 for more info on why, and let me know if it solves your question.

**Comment by yjiakang** [comment]
I read through the discussion in #187,  and the taxonomic and functional annotation parts. Now, I am clear. Thanks.

**Comment by yjiakang** [comment]
sqm_reads.pl stucked when running, here is the output. And it is still running. Should I be more patient or sth is wrong?
```
Setting block size for Diamond
  AVAILABLE (free) RAM memory: 93.76 Gb
We will set Diamond block size to 8 (Gb RAM/5, Max 8). You can override this setting using the -b option when starting the project.

Now reading samples from /public/home/jkyin/yjk/tomato_micro/sqm_one_by_one/sample_file/A57RS1T1
1 metagenomes found

SAMPLE 1/1: A57RS1T1

   File: A57RS1T1_1.fa
[38 seconds]: Running Diamond (Buchfink et al 2015, Nat Methods 12, 59-60) for taxa (GenBank nr, Clark et al 2016, Nucleic Acids Res 44, D67-D72)
[2 days, 0 hours, 25 minutes, 19 seconds]: Running LCA
  Splitting Diamond file
  Starting multithread LCA in 8 threads: 1 2 3 4 5 6 7 8
```

**Comment by jtamames** [comment]
Hello
Apparently, you provided a very big dataset (Diamond took 2 days to complete), so I would expect the LCA taking also a long time. Did you check if the wranks file in the A57RS1T1 directory  is growing? If it is, then just patience is required

Best,
J

**Comment by yjiakang** [comment]
Yeah, the wranks file is growing, I will be more patient. 
Best
Jk Yin

**Comment by yjiakang** [comment]
I used 8 cores to run the sqm_read.pl for just one sample, and the process lasted 20 days, so I gived up for the economic and project cycle reasons. And I think more samples would need more time and computational resources. Alternatively, I choosed some samples that is computationally acceptable to run in coassembly mode. And it runs smoothly. Anyway, thanks for your pipeline and your help with regard to all my questions.
Best
